aloha-core/src/main/scala/com/eharmony/aloha/models/multilabel/MultilabelModel.scala

```diff
@@ -0,0 +1,495 @@
+    * @param missing missing features from
+    * @param prediction the prediction(s) made by the embedded predictor.
+    * @param auditor an auditor used to audit the output.
+    * @tparam U upper bound on model output type `B`
+    * @tparam K type of label or class
+    * @tparam B output type of the model.
+    * @return a SubValue indicating success.
+    */
+  protected[multilabel] def reportSuccess[U, K, B <: U](
+      modelId: ModelIdentity,
+      labelInfo: LabelsAndInfo[K],
+      missing: scm.Map[String, Seq[String]],
+      prediction: Map[K, Double],
+      auditor: Auditor[U, Map[K, Double], B]
+  ): Subvalue[B, Map[K, Double]] = {
+
+    val aud = auditor.success(
+      modelId,
+      prediction,
+      errorMsgs = labelInfo.errorMsgs,
+      missingVarNames = combineMissing(labelInfo, missing)
+    )
+
+    Subvalue(aud, Option(prediction))
+  }
+
+  /**
+    * Report that a `Throwable` was thrown while invoking the predictor
+    * @param modelId An identifier for the model.  Used in error reporting.
+    * @param labelInfo labels and information about the labels.
+    * @param missingFeatureMap missing features from RegressionFeatures
+    * @param throwable the error the occurred in the predictor.
+    * @param auditor an auditor used to audit the output.
+    * @tparam U upper bound on model output type `B`
+    * @tparam K type of label or class
+    * @tparam B output type of the model.
+    * @return a SubValue indicating failure.
+    */
+  protected[multilabel] def reportPredictorError[U, K, B <: U](
+      modelId: ModelIdentity,
+      labelInfo: LabelsAndInfo[K],
+      missingFeatureMap: scm.Map[String, Seq[String]],
+      throwable: Throwable,
+      auditor: Auditor[U, Map[K, Double], B]
+  ): Subvalue[B, Nothing] = {
+
+    val sw = new StringWriter
+    val pw = new PrintWriter(sw)
+    throwable.printStackTrace(pw)
+    val stackTrace = sw.toString.split("\n").take(NumLinesToKeepInStackTrace).mkString("\n")
+
+    val aud = auditor.failure(
+      modelId,
+      errorMsgs = stackTrace +: labelInfo.errorMsgs,
+      missingVarNames = combineMissing(labelInfo, missingFeatureMap)
+    )
+    Subvalue(aud, None)
+  }
+
+  /**
+    * Get labels from the input for which a prediction should be produced.
+    * @param example the example provided to the model
+    * @param labelsOfInterest a function used to extract labels for which a
+    *                         prediction should be produced.
+    * @param labelToInd mapping from Label to index into the sequence of all
+    *                   labels seen in the training set.
+    * @tparam A input type of the model
+    * @tparam K type of label or class
+    * @return labels and information about the labels.
+    */
+  protected[multilabel] def labelsForPrediction[A, K](
+      example: A,
+      labelsOfInterest: GenAggFunc[A, sci.IndexedSeq[K]],
+      labelToInd: Map[K, Int]
+  ): LabelsAndInfo[K] = {
```

Comment:
How often is this function called?  It seems like it can be written more efficiently if it's hot.

Reply:
Once per prediction.  How would you speed it up?